llvm/lib/ExecutionEngine/SectionMemoryManager.cpp

```diff
@@ -18,6 +18,89 @@
 
 namespace llvm {
 
+bool SectionMemoryManager::hasSpace(const MemoryGroup &MemGroup,
+                                    uintptr_t Size) const {
+  for (const FreeMemBlock &FreeMB : MemGroup.FreeMem) {
+    if (FreeMB.Free.allocatedSize() >= Size)
+      return true;
+  }
+  return false;
+}
+
+void SectionMemoryManager::reserveAllocationSpace(
+    uintptr_t CodeSize, Align CodeAlign, uintptr_t RODataSize,
+    Align RODataAlign, uintptr_t RWDataSize, Align RWDataAlign) {
+  if (CodeSize == 0 && RODataSize == 0 && RWDataSize == 0)
```

Comment:
Perhaps we should clear the `FreeMem` vectors before returning early here? Otherwise, we could reserve no space, but still have some free memory left over from a previous reservation, then end up using that to satisfy a later erroneous request instead of hitting the assert that fails when we try to allocate memory without having reserved it first.

Reply:
Or, is this OK to return early without clearing free memory, on the same grounds that we reuse a previous contiguous block [below](https://github.com/llvm/llvm-project/pull/71968/files#diff-58b356a4f34602c02eed8c5a40839ebbc5149fefa38b6739a87c9b85862bf372R49-R50) if there's enough space in the existing reservation?

Reply:
Erroneous requests would negate any benefits from using reserveAllocationSpace. And yes, if I removed this early return it would still return early when checking for sufficient existing free space to satisfy the whole request.